### PR TITLE
[codex] fix linux AppImage launcher startup

### PR DESCRIPTION
**Pull request by hashimkarim**
## Why

Linux AppImage launches can inherit `ELECTRON_RUN_AS_NODE=1` from Electron-based shells, which makes the Electron binary behave like Node instead of starting the desktop app. The generated AppImage entry also tried to `require()` an ESM-only package, which fails during packaged startup.

This follow-up also keeps tools-pack lifecycle ownership intact for direct `$APPDIR/AppRun` launches: when `$APPIMAGE` is missing, the wrapper exports `APPIMAGE=$APPDIR/AppRun`, and `tools-pack linux stop` / `uninstall` must still recognize the running desktop process as ours.

## What users will see

Linux AppImage desktop/menu launches start the packaged Open Design app normally instead of starting as Node. Direct `AppRun` launches remain manageable by `tools-pack linux stop`, `uninstall`, and cleanup.

## Surface area

- [ ] **UI** — new page / dialog / panel / menu item / setting / empty state in `apps/web` or `apps/desktop` (including Electron menu bar)
- [ ] **Keyboard shortcut** — new or changed
- [ ] **CLI / env var** — new `od` subcommand or flag, new `tools-dev` / `tools-pack` / `tools-pr` flag, or new `OD_*` env var
- [ ] **API / contract** — new `/api/*` endpoint, new SSE event, or changed shape in `packages/contracts`
- [ ] **Extension point** — new entry under `skills/`, `design-systems/`, `design-templates/`, or `craft/`, or change to the skills protocol
- [ ] **i18n keys** — added new translation keys (see `TRANSLATIONS.md` for the locale workflow)
- [ ] **New top-level dependency** — adding any new entry to the **root** `package.json` (`dependencies` or `devDependencies`); workspace-package `package.json` files are out of scope. Include a paragraph on what we get vs. what bytes we ship (see `CONTRIBUTING.md` → Code style)
- [ ] **Default behavior change** — changes what existing users experience without opting in (default model, default setting, file/SQLite schema, auto-network on startup, auto-install)
- [x] **None** — internal refactor, docs, tests, or translation update only

## Screenshots

N/A — packaging/lifecycle-only fix.

## Bug fix verification

- Test path that reproduces the lifecycle bug: `tools/pack/tests/linux.test.ts`
- Before: AppImage launches could inherit `ELECTRON_RUN_AS_NODE=1` and start as Node; direct `$APPDIR/AppRun` launches set `APPIMAGE=$APPDIR/AppRun` only as a shell-local variable, so the execed Electron process could still miss `APPIMAGE` and be misclassified as unmanaged by lifecycle validation.
- Now: the AppRun wrapper unsets `ELECTRON_RUN_AS_NODE` before exec and exports `APPIMAGE=$APPDIR/AppRun`; IPC verification reports `state: running`; and the direct-AppRun regression tests verify both the wrapper-exported child environment and `stopPackedLinuxApp()` ownership.
- Red/green: not rerun on `main` during this takeover pass; this branch adds regression coverage for the reviewer-flagged lifecycle shapes.

## Validation

- `git diff --check`
- `pnpm --filter @open-design/tools-pack test -- tests/linux.test.ts`
- `pnpm --filter @open-design/tools-pack build`
- built Linux AppImage under Node 24.14.1
- relaunched the integrated AppImage locally and verified desktop IPC reports `state: running`, `title: Open Design`, `url: od://app/`
- `pnpm -C tools/pack test linux.test.ts` (this takeover pass; passed 61 tests, with local Node engine warning for v26 vs expected ~24)

## Follow-up

- #4642 fixes the develop-menu shortcut dialog observed during Linux AppImage testing. On affected sessions, startup paused until the user dismissed `Develop menu shortcut unavailable`, even though the app could continue normally.

### CODE REVIEW

**Comment by lefarcen** [comment]
Hey @Hashim-K! 👋 The AppImage startup fix is aimed at a real packaged Linux launch path, which is the right direction.

Thanks for opening this as a draft — once you mark it ready for review, the reviewer pool can take a full pass. Until then I'll stay out of the way to avoid noise during iteration.

**Comment by lefarcen** [comment]
Hey @Hashim-K — one concrete review note has landed on the current draft: Copilot flagged the new keep-alive timer in `tools/pack/src/linux.ts` as something worth tightening so it doesn't keep the launcher alive longer than needed on early-exit paths.

Since this PR is still in draft, I'd treat that as a good thing to fold in during the next pass, then the reviewer pool can look at the updated head once you mark it ready.

File: tools/pack/src/linux.ts [hunk 1]
```diff
@@ -505,7 +505,7 @@ async function writeAssembledApp(
   };
   await writeFile(paths.assembledPackageJsonPath, `${JSON.stringify(packageJson, null, 2)}\n`, "utf8");
 
-  const mainStub = `"use strict";\nrequire("@open-design/packaged");\n`;
+  const mainStub = `"use strict";\n// Keep Electron's main process alive while the ESM packaged entry imports and\n// reaches app.whenReady(). Dynamic import alone is not a libuv handle.\nsetTimeout(() => {}, 60_000);\nimport("@open-design/packaged").catch((error) => {\n  console.error("packaged entry failed", error);\n  process.exit(1);\n});\n`;
   await writeFile(paths.assembledMainEntryPath, mainStub, "utf8");
```
[the remark this answers is not in the document]
Comment (Hashim-K) [reply tools/pack/src/linux.ts]: Fixed in a4e5ae18. The generated Linux `main.cjs` no longer uses the temporary keep-alive timer; it now uses a plain dynamic `import("@open-design/packaged")` with a catch handler, so there is no timer left to keep the launcher alive on early-exit paths.

File: tools/pack/src/linux.ts [hunk 2]
```diff
@@ -260,6 +260,61 @@ export function renderDesktopTemplate(template: string, values: DesktopTemplateV
...
+    if [ $NUMBER_OF_ARGS -eq 0 ] ; then
+      exec "$BIN"
+    else
+      exec "$BIN" "\${args[@]}"
+    fi
+  fi
+}
+
+if [ -z "$APPIMAGE" ] ; then
+  APPIMAGE="$APPDIR/AppRun"
```
Comment (nettee) [inline tools/pack/src/linux.ts]: This fallback rewrites `$APPIMAGE` to `"$APPDIR/AppRun"`, but our own lifecycle validation still treats extracted/FUSE launches as owned only when `$APPIMAGE` points back to the installed `.AppImage` path. `matchesAppImageProcess()` makes that assumption at `tools/pack/src/linux.ts:329-336`, and `validateDesktopAppImageMarker()` uses it to decide whether `tools-pack linux stop` / `uninstall` can manage a packaged desktop launch at `tools/pack/src/linux.ts:991-1033`. In the exact direct-`AppRun` mode this branch is handling, the process will now inherit `APPIMAGE=/.../AppRun` instead, so menu-launched apps can be misclassified as `unmanaged` or `not-running` even though they are ours. Please either preserve the original AppImage path here when it is known, or extend the validator to accept this direct-`AppRun` shape before landing the wrapper.
<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found one blocking lifecycle regression in the new Linux AppImage wrapper path; details are inline.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
@Hashim-K — nettee's review above flags a blocking lifecycle issue at \`linux.ts:310\`: the \`\$APPIMAGE → \$APPDIR/AppRun\` rewrite can cause \`matchesAppImageProcess()\` to misclassify a running app as unmanaged. Worth addressing that inline comment before the next review pass.

Also two template items still open: **Surface area** checkbox (likely \`None\` for this packaging-only fix) and a short **Bug fix verification** (e.g. "was: AppImage inherits ELECTRON_RUN_AS_NODE=1, starts as Node; now: IPC confirms \`state: running\`"). Your Validation section has the commands — just need the explicit before/after seam.

💡 To drive this to merge hands-free, paste this to your AI coding agent (Claude Code / Codex / opencode / Cursor …):
`Take over nexu-io/open-design#4494 until it merges — read https://raw.githubusercontent.com/nexu-io/looper/main/skills/pr-takeover/SKILL.md and follow it.`

File: tools/pack/src/linux.ts (hunk 2 above)
Comment (Hashim-K) [reply tools/pack/src/linux.ts]: Fixed in 55039af4: `matchesAppImageProcess()` now accepts the direct-AppRun shape only when the live Electron executable is `Open Design` and `$APPIMAGE` points to that executable directory sibling `AppRun`. That lets `validateDesktopAppImageMarker()` keep menu/direct-AppRun launches owned instead of returning unmanaged, while the existing stamp + namespace checks still gate lifecycle ownership. Added matcher coverage and a `stopPackedLinuxApp()` regression test with a live direct-AppRun-shaped process.

**Comment by Hashim-K** [comment]
@nettee addressed the lifecycle blocker in 55039af4 and resolved the inline thread. Summary: direct AppRun launches are now accepted only when the live `Open Design` executable has `$APPIMAGE` pointing to its sibling `AppRun`, while the existing stamp and namespace checks still gate ownership. Also filled the Surface area and Bug fix verification template items. Could you re-review when you get a chance?

File: tools/pack/src/linux.ts (hunk 2 above)
Comment (nettee) [inline tools/pack/src/linux.ts]: When `$APPIMAGE` is missing, this assigns `APPIMAGE="$APPDIR/AppRun"` only as a shell-local variable. The later `exec "$BIN"` does not inherit non-exported assignments, so the Electron process still starts without `APPIMAGE` in the real direct-`AppRun` path. That means the new `matchesAppImageProcess()` fallback at `tools/pack/src/linux.ts:339-344` will still reject the process during `linux stop` / `uninstall`, even though this PR is trying to preserve that ownership shape. Please export the fallback value here (for example `export APPIMAGE="$APPDIR/AppRun"`) and add a regression test that executes the generated wrapper and asserts the child process environment actually contains `APPIMAGE`, not just that the script text contains the assignment.
<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found one blocking regression in the direct-AppRun lifecycle path; details are inline.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: tools/pack/src/linux.ts (hunk 2 above)
Comment (Hashim-K) [reply tools/pack/src/linux.ts]: Fixed in 592b5894: the AppRun fallback now uses `export APPIMAGE="$APPDIR/AppRun"`, so the execed Electron process inherits the marker. I also added a regression test that runs the generated wrapper against a fake `Open Design` binary and asserts the child receives `APPIMAGE=<appdir>/AppRun` while `ELECTRON_RUN_AS_NODE` is unset.

**Review by nettee — APPROVED** [review]
@Hashim-K I rechecked the current Linux AppImage launcher changes on `592b5894d02cbab221bff1064752e54292febcf0`: the packaged entry now loads the ESM runtime without `require()`, both the desktop template and `tools-pack linux start` clear `ELECTRON_RUN_AS_NODE`, and the custom AppRun wrapper now exports the fallback `APPIMAGE` marker so direct `AppRun` launches stay owned by the lifecycle validation path. The added tests cover the desktop template, the spawn environment, the direct-AppRun ownership matcher, and the wrapper's child-process env propagation. Thanks for tightening these Linux lifecycle edge cases and following through on the reviewer feedback.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by Hashim-K** [comment]
Approved and all visible checks are green on 592b5894. I attempted to merge, but the repo has merge queue enabled and this account lacks permission for `EnablePullRequestAutoMerge`, so GitHub will not let me enqueue/merge it from here. Could a maintainer with merge-queue permission enqueue or merge this PR?
